src/java/org/apache/cassandra/repair/messages/RepairMessage.java

```diff
@@ -51,48 +65,165 @@ protected RepairMessage(RepairJobDesc desc)
+    {
+        RepairRetrySpec.Verb configVerb = toConfigVerb(verb);
+        if (configVerb == null)
+            return Backoff.None.INSTANCE;
+        RepairRetrySpec retrySpec = DatabaseDescriptor.getRepairRetrys();
+        RetrySpec spec = retrySpec.get(configVerb);
+        if (!spec.isEnabled())
+            return Backoff.None.INSTANCE;
+        switch (spec.type)
+        {
+            case Expoential:
+                return new Backoff.ExpoentialBackoff(spec.maxAttempts.value, spec.baseSleepTime.toMilliseconds(), spec.maxSleepTime.toMilliseconds(), ctx.random().get()::nextDouble);
+            default:
+                throw new IllegalArgumentException("Unknown type: " + spec.type);
+        }
+    }
+
+    @Nullable
+    private static RepairRetrySpec.Verb toConfigVerb(Verb verb)
     {
-        RequestCallback<?> callback = new RequestCallback<Object>()
+        switch (verb)
+        {
+            case PREPARE_MSG: return RepairRetrySpec.Verb.PREPARE;
+            case VALIDATION_REQ: return RepairRetrySpec.Verb.VALIDATION_REQ;
+            case VALIDATION_RSP: return RepairRetrySpec.Verb.VALIDATION_RSP;
+            case SYNC_REQ: return RepairRetrySpec.Verb.SYNC_REQ;
+            case SYNC_RSP: return RepairRetrySpec.Verb.SYNC_RSP;
+            case SNAPSHOT_MSG: return RepairRetrySpec.Verb.SNAPSHOT;
+            case CLEANUP_MSG: return RepairRetrySpec.Verb.CLEANUP;
+            default: return null;
+        }
+    }
+
+    public static <T> void sendMessageWithRetries(SharedContext.MessagingContext ctx, RepairMessage request, Verb verb, InetAddressAndPort endpoint, RequestCallback<T> finalCallback)
+    {
+        // 0 means no retries
+        sendMessageWithRetries(ctx, backoff(ctx, verb), request, verb, endpoint, finalCallback, 0);
+    }
+
+    public static <T> void sendMessageWithRetries(SharedContext.MessagingContext ctx, RepairMessage request, Verb verb, InetAddressAndPort endpoint)
+    {
+        sendMessageWithRetries(ctx, backoff(ctx, verb), request, verb, endpoint, new RequestCallback<>()
         {
             @Override
             public void onResponse(Message<Object> msg)
             {
-                logger.info("[#{}] {} received by {}", request.desc.parentSessionId, verb, endpoint);
-                // todo: at some point we should make repair messages follow the normal path, actually using this
+                // TODO (now): define
+            }
+
+            @Override
+            public void onFailure(InetAddressAndPort from, RequestFailureReason failureReason)
+            {
+                // TODO (now): define
+            }
+        }, 0);
+    }
+
+    private static <T> void sendMessageWithRetries(SharedContext.MessagingContext ctx, Backoff backoff, RepairMessage request, Verb verb, InetAddressAndPort endpoint, RequestCallback<T> finalCallback, int attempt)
+    {
+        // TODO (now, api): expose a way for callers to stop early even if backoff allows more progress
```

Comment:
The idea would be to prevent further retry attempts in `onFailure()`? I suppose we could do that if we added short-circuit/stop request to the contract of `RequestCallback` and returned it from this method?

Reply:
correct, short-circuit if we no longer care